cats/cats-dynomite/src/main/java/com/netflix/spinnaker/cats/dynomite/cluster/DynoClusteredAgentScheduler.java

```diff
@@ -0,0 +1,217 @@
+
+  public DynoClusteredAgentScheduler(DynomiteClientDelegate redisClientDelegate, NodeIdentity nodeIdentity, AgentIntervalProvider intervalProvider, NodeStatusProvider nodeStatusProvider) {
+    this(redisClientDelegate, nodeIdentity, intervalProvider, nodeStatusProvider, Executors.newSingleThreadScheduledExecutor(new NamedThreadFactory(ClusteredAgentScheduler.class.getSimpleName())), Executors.newCachedThreadPool(new NamedThreadFactory(AgentExecutionAction.class.getSimpleName())));
+  }
+
+  public DynoClusteredAgentScheduler(DynomiteClientDelegate redisClientDelegate, NodeIdentity nodeIdentity, AgentIntervalProvider intervalProvider, NodeStatusProvider nodeStatusProvider, ScheduledExecutorService lockPollingScheduler, ExecutorService agentExecutionPool) {
+    this.redisClientDelegate = redisClientDelegate;
+    this.nodeIdentity = nodeIdentity;
+    this.intervalProvider = intervalProvider;
+    this.nodeStatusProvider = nodeStatusProvider;
+    this.agentExecutionPool = agentExecutionPool;
+    lockPollingScheduler.scheduleAtFixedRate(this, 0, 5, TimeUnit.SECONDS);
+  }
+
+  @Override
+  public void schedule(Agent agent, AgentExecution agentExecution, ExecutionInstrumentation executionInstrumentation) {
+    if (agent instanceof AgentSchedulerAware) {
+      ((AgentSchedulerAware) agent).setAgentScheduler(this);
+    }
+
+    final AgentExecutionAction agentExecutionAction = new AgentExecutionAction(agent, agentExecution, executionInstrumentation);
+    agents.put(agent.getAgentType(), agentExecutionAction);
+  }
+
+  @Override
+  public void unschedule(Agent agent) {
+    releaseRunKey(agent.getAgentType(), 0);
+    agents.remove(agent.getAgentType());
+  }
+
+  @Override
+  public void run() {
+    if (!nodeStatusProvider.isNodeEnabled()) {
+      return;
+    }
+    try {
+      runAgents();
+    } catch (Throwable t) {
+      t.printStackTrace(System.err);
+    }
+  }
+
+  private Map<String, Long> acquire() {
+    Map<String, Long> acquired = new HashMap<>(agents.size());
+    Set<String> skip = new HashSet<>(activeAgents.keySet());
+    agents.entrySet().stream()
+      .filter(a -> !skip.contains(a.getKey()))
+      .forEach(a -> {
+        final String agentType = a.getKey();
+        AgentIntervalProvider.Interval interval = intervalProvider.getInterval(a.getValue().getAgent());
+        if (acquireRunKey(agentType, interval.getTimeout())) {
+          acquired.put(agentType, System.currentTimeMillis() + interval.getInterval());
+        }
+      });
+    return acquired;
+  }
+
+  private boolean acquireRunKey(String agentType, long timeout) {
+    // This isn't safe, but it's the best one can do without `String set(String key, String value, String nxxx, String expx, long time).`
+    return redisClientDelegate.withCommandsClient(client -> {
+      String response = client.get(agentType);
+      if (response == null) {
+        if (client.setnx(agentType, nodeIdentity.getNodeIdentity()) == 1) {
```

Comment:
That's right, there's a bit of sneakiness that can happen, in addition to the consistency thing I mentioned above. CDE recommends moving things like this, as well as the task repository stuff into its own cluster so it can run in quorum. However, once Lua is added (coming this quarter), we can switch to the other agent and we'll hash to the same server, having the same atomic guarantees for locks as we have in Redis today.